src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -556,6 +556,227 @@ inline bool ReduceAxesOpForwardStorage(const nnvm::NodeAttrs& attrs,
+  mshadow::Tensor<xpu, 1, uint8_t> workspace =
+    ctx.requested[0].get_space_typed<xpu, 1, uint8_t>(
+      mshadow::Shape1(memSize), ctx.get_stream<xpu>());
+  return reinterpret_cast<DType *>(workspace.dptr_);
+}
+
+template<typename xpu, typename DType>
+void ArgMax(const OpContext& ctx, const TShape &shape, int axis, size_t step, size_t shift,
+            const int nWorkers, void *pIdxStorage, const idxStorageType storageType,
+            const TBlob& input, const TBlob& output) {
+  using namespace mxnet_op;
+  const auto pIn = input.dptr<DType>();
+  auto *pOut = output.dptr<DType>();
+  const auto nSize = shape[axis];
+  const auto num_threads = shape.Size() / nSize;
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  Kernel<argmax, xpu>::Launch(s, num_threads * nWorkers, nWorkers, pIn, pOut, nSize,
+                              step, shift, pIdxStorage, storageType);
+  if (nWorkers > 1) {
+    Kernel<argmax_reduce, xpu>::Launch(s, num_threads, nWorkers, pIn, pOut,
+                                       step, shift, pIdxStorage, storageType);
+  }
+}
+
+template<typename xpu>
+int DefineNumbWorkers(const TShape &shape, int axis);
+
+template<typename xpu>
+void ArgMax(const nnvm::NodeAttrs& attrs,
+            const OpContext& ctx,
+            const std::vector<TBlob>& inputs,
+            const std::vector<OpReqType>& req,
+            const std::vector<TBlob>& outputs) {
+  const ReduceAxisParam& param = nnvm::get<ReduceAxisParam>(attrs.parsed);
+  auto shape = inputs[0].shape_;
+  size_t shift = shape.Size();     // to reading data starting of each thread
+  CHECK_NE(shift, 0U) << "attempt to search an empty sequence";
+
+  size_t step = 1;  // step over the input array which will be used by each thread in kernel
+  int axis;
+  if (param.axis.has_value()) {
+    axis = CheckAxis(param.axis.value(), shape.ndim());
+    // cannot do argmax in an empty dimension
+    CHECK_NE(shape[axis], 0)
+      << "searching input tensor of shape " << shape
+      << " along axis = " << axis << " of zero dim-size is not allowed";
+
+    if (shape.ndim() == 1)
+      shape = AxisShapeCompact(shape, &axis, true);
+
+    // Calculate step & shift
+    auto i = shape.ndim();
+    while (--i > axis)
+      step *= shape[i];
+
+    shift = i? step*shape[i] : 1;
+  } else {
+    // If global reduction, reshape the input tensor into 2D shape (1, inputs[0].shape_.Size())
+    // and search on axis = 1.
+    shape = mxnet::TShape(2, 1);
+    shape[1] = shift;
+    axis = 1;
+  }
+
+  void *pIdxMemory = nullptr;
+  auto nWorkers = DefineNumbWorkers<xpu>(shape, axis);
+  idxStorageType storageType = undefined_st;
+  while (nWorkers > 1) {
+    const size_t num_items = nWorkers * shape.Size() / shape[axis];
+#if MXNET_USE_INT64_TENSOR_SIZE
+    pIdxMemory = AllocateDTypeMemory<xpu, index_t>(ctx, num_items);
+    if (pIdxMemory) {
+      storageType = int64_st;
+      break;
+    }
+#endif
+    if (shape[axis] <= UINT32_MAX) {
```

Comment:
I think `uint32_t` is not being used for shape index any more. It's either `int64_t` or `int32_t` depending on the compiler flag. As I commented earlier, it might be cleaner to just pass in `index_t` directly.

Reply:
Actually, in my initial version (when the use of `int64_t` was not implemented):
```
 auto nWorkers = DefineNumbWorkers<xpu>(shape, axis);
 bool use_uin16 = false;
  while (nWorkers > 1) {
    const size_t num_items = nWorkers * shape.Size() / shape[axis];
      pIdxMemory = AllocateDTypeMemory<xpu, uint32_t>(ctx, num_items);
      if (pIdxMemory)
        break;

    // Check if indexes can be stored in uint16 format
    if (shape[axis] <= UINT16_MAX) {
      pIdxMemory = AllocateDTypeMemory<xpu, uint16_t>(ctx, num_items);
      if (pIdxMemory) {
        use_uin16 = true;
        break;
      }
    }

    // Decreasing the number of workers - less memory will be needed
    nWorkers >>= 1;
  }
```
I was getting "optimal" value of `nWorkers` first, and I tried to allocate memory for indexes in "unsigned" formats.  Perhaps, it is not always true (and, please, correct me, if I am wrong), but I noticed that with `uint32_t` pointers the program runs 5-10% faster than with `uint16_t`. Because of that, I was trying to allocate memory in that order. 
When 
- (a) for some value of `nWorkers` the program fails to allocate `uint32_t *` memory AND
- (b) `shape[axis] <= UINT16_MAX`, therefore it is still possible to use `uint16_t *` memory, 

I was trying to allocate and use `uint16_t *` memory for the same value of `nWorkers`. But, if
- (a) is not possible to use "shorter" memory (`shape[axis] > UINT16_MAX)` OR 
- (b) the attempt to allocate it failed, 

I reduced nWorkers >> = 1 and started a new round of memory allocation with smaller `nWorkers`. (Yes, now `nWorkers` is not optimal, but it's better to as much as possible workers!).

After I started to use `index_t`, I tried to follow the same pattern.

**NOTE,** when `index_t` is defined as `int64_t` I also see 5-10% run time degradation, when 
`int64_t *` memory is used for indexes storage. Therefore, with regard to runtime, this is not the best solution, but at least there is some logic here. Best solution would be use `int64_t` pointers **if and only if** it is impossible to store indexes in `uint32_t` format, otherwise use 
`uint32_t` or `uint_16_t` formats.

Definitely, I could forget about optimization of run time and use **only** the pointers of type `index_t *`. In that case the loop will be quite simple:
```
  while (nWorkers > 1) {
    const size_t num_items = nWorkers * shape.Size() / shape[axis];
    pIdxMemory = AllocateDTypeMemory<xpu, index_t>(ctx, num_items);
    if (pIdxMemory)
       break;
    nWorkers >>= 1;
  }
```

So it looks like we have following three options:
(1) keep current implementation (which looks logical, but note the best one when `index_t` is defined as `int64_t`)
(2) store indexes only `index_t` format (which is the simplest, but not the best one with respect to run time)
(3) implement the **"best"** solution: use `int64_t ` format only when `index_t` is defined as 'int64_t' AND `shape[axis] > UINT32_MAX`  

Which one do you prefer?